core/codec/json/basic_coding.hpp

```diff
@@ -0,0 +1,285 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <rapidjson/document.h>
+
+#include <array>
+#include <boost/optional.hpp>
+#include <cppcodec/base64_rfc4648.hpp>
+#include <gsl/span>
+#include <map>
+#include <set>
+#include <tuple>
+#include <vector>
+
+#include "codec/json/json_errors.hpp"
+#include "common/outcome.hpp"
+
+#define COMMA ,
+
+#define JSON_ENCODE(type)               \
+  inline fc::codec::json::Value encode( \
+      const type &v, rapidjson::MemoryPoolAllocator<> &allocator)
+
+// NOLINTNEXTLINE(bugprone-macro-parentheses)
+#define JSON_DECODE(type) \
+  inline void decode(type &v, const fc::codec::json::Value &j)
+
+namespace fc::codec::json {
+  using rapidjson::Document;
+  using rapidjson::Value;
+  using base64 = cppcodec::base64_rfc4648;
+
+  template <typename T>
+  T innerDecode(const Value &j);
+
+  template <typename T>
+  void Set(Value &j,
+           std::string_view key,
+           const T &v,
+           rapidjson::MemoryPoolAllocator<> &allocator);
+
+  template <typename T>
+  inline T kDefaultT() {
+    return {};
+  }
```

Comment:
`kDefault` used by cbor should be reused to avoid duplication? Can be moved/renamed